[#2053] Adding integration tests for gateway-based auto provisioning.

fkaltner opened the pull request:
This PR adds integration tests for auto-provisioning.

Before merging this PR 
https://github.com/eclipse/hono/pull/2358
https://github.com/eclipse/hono/pull/2360
https://github.com/eclipse/hono/pull/2359
need to be merged, so that tests will run successfully

sophokles73 commented on tests/src/test/java/org/eclipse/hono/tests/amqp/AmqpUploadTestBase.java:
```diff
@@ -241,6 +244,76 @@ public void testAdapterClosesLinkOnMessageExceedingMaxPayloadSize(final VertxTes
+
+        final String deviceId = helper.getRandomDeviceId(tenantId);
+        msg.setAddress(String.format("%s/%s/%s", getEndpointName(), tenantId, deviceId));
+        sender.send(msg, delivery -> {
```
you should either consider the senderQoS being passed into the test method or do not run this as a parameterized test. FMPOV the latter would be ok.

sophokles73 commented on tests/src/test/java/org/eclipse/hono/tests/amqp/AmqpUploadTestBase.java:
```diff
@@ -241,6 +244,76 @@ public void testAdapterClosesLinkOnMessageExceedingMaxPayloadSize(final VertxTes
+            messageSent.completeNow();
+        });
+
+        assertThat(messageSent.awaitCompletion(5, TimeUnit.SECONDS)).isTrue();
+        assertThat(emptyEventReceived.awaitCompletion(5, TimeUnit.SECONDS)).isTrue();
+        assertThat(telemetryReceived.awaitCompletion(5, TimeUnit.SECONDS)).isTrue();
```
how about using a simple Checkpoint instead of coordinating three TestContexts?

sophokles73 commented on tests/src/test/java/org/eclipse/hono/tests/coap/CoapTestBase.java:
```diff
@@ -408,6 +410,56 @@ public void testUploadMessagesViaGateway(final VertxTestContext ctx) throws Inte
+    @Test
+    public void testAutoProvisioningViaGateway(final VertxTestContext ctx) throws InterruptedException {
+
+        // GIVEN a device that is connected via two gateways
```
one gateway, isn't it?

sophokles73 commented on tests/src/test/java/org/eclipse/hono/tests/coap/CoapTestBase.java:
```diff
@@ -408,6 +410,56 @@ public void testUploadMessagesViaGateway(final VertxTestContext ctx) throws Inte
+                }),
+                close -> {});
+
+        testUploadMessages(ctx, tenantId,
```
IMHO sending the message could/should be done in a much simpler way, e.g. as in the *testUploadMessageFailsForDisabledDevice* test method ...

sophokles73 commented on tests/src/test/java/org/eclipse/hono/tests/http/HttpTestBase.java:
```diff
@@ -857,6 +859,64 @@ public void testUploadMessageFailsForUnauthorizedGateway(final VertxTestContext
+            });
+        }, remoteClose -> {});
+
+        testUploadMessages(
```
same here, simply send one request and handle its outcome directly ...

fkaltner commented:
I reworked the tests for AMQP, CoAP and MQTT. I think this addresses your comments so far.

Nevertheless I am not yet done, since I would like to take rework the MQTT test as well and remove the unnecessary changes I made because I sent multiple messages instead of a single one...

fkaltner commented:
Done! MQTT test is now refactored too 😎

sophokles73 reviewed (approved):
LGTM
